src/cli.rs

```diff
@@ -585,12 +588,43 @@ async fn run_deploy(
         )
     };
 
-    for data in data_iter() {
-        deploy::push::build_profile(data).await?;
-    }
+    let (remote_builds, local_builds): (Vec<_>, Vec<_>) = data_iter().partition(|data| {
+        data.deploy_data.merged_settings.remote_build.unwrap_or_default()
+    });
+
+    // the grouping by host will retain each hosts ordering by profiles_order since the fold is synchronous
+    let remote_build_map: HashMap<_, Vec<_>> = remote_builds.iter().fold(HashMap::new(), |mut accum, elem| {
+        match accum.get_mut(elem.deploy_data.node_name) {
+            Some(v) => { v.push(elem); accum },
+            None => { accum.insert(elem.deploy_data.node_name, vec![elem]); accum }
+        }
+    });
+
+    let remote_results = join!(
+        // remote builds can be run asynchronously (per host)
+        join_all(remote_build_map.into_iter().map(deploy_profiles_to_host)),
+        async {
+            // run local builds synchronously to prevent hardware deadlocks
+            for data in &local_builds {
+                deploy::push::build_profile(data).await.unwrap();
```

Comment:
I'm not a huge fan of using "partial functions", isn't unhandled panic from `unwrap` going to kill the main thread?
Also, AFAICS, at the moment we completely ignore non-remote build/push results